compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SemanticAnalyzer.java

```diff
@@ -1373,17 +1376,42 @@ private boolean validateErrorVariable(BLangErrorVariable errorVariable, BErrorTy
         }
 
         if (isRestDetailBindingAvailable(errorVariable)) {
+            // Type of rest pattern is a map type where constraint type is,
+            // union of keys whose values are not matched in error binding/match pattern.
             BTypeSymbol typeSymbol = createTypeSymbol(SymTag.TYPE);
-            BMapType restType = new BMapType(TypeTags.MAP, recordType.restFieldType, typeSymbol);
+            BType constraint = getRestMapConstraintType(detailFields, matchedDetailFields, recordType);
+            BMapType restType = new BMapType(TypeTags.MAP, constraint, typeSymbol);
             typeSymbol.type = restType;
             errorVariable.restDetail.type = restType;
             errorVariable.restDetail.accept(this);
         }
         return true;
     }
 
-    // todo: warn parameter is a hack to fix patch compatibility by using dlog.warn, remove for minor release
-    private boolean validateErrorReasonMatchPatternSyntax(BLangErrorVariable errorVariable, boolean warn) {
+    private BType getRestMapConstraintType(Map<String, BField> errorDetailFields, Set<String> matchedDetailFields,
+                                           BRecordType recordType) {
+        BUnionType restUnionType = BUnionType.create(null);
+        if (!recordType.sealed) {
+            restUnionType.add(recordType.restFieldType);
+        }
+        for (Map.Entry<String, BField> entry : errorDetailFields.entrySet()) {
+            if (!matchedDetailFields.contains(entry.getKey())) {
+                BType type = entry.getValue().getType();
+                if (!types.isAssignable(type, restUnionType)) {
```

Comment:
Shouldn't we check this both ways to extract out the narrowest type like in https://github.com/ballerina-platform/ballerina-lang/pull/20180/files#diff-782094a79e6831ac57a34ad157af49c1R1178-R1186?

For example if the rest type is a finite type Foo

```ballerina
type Foo "foo"|"bar";
```

And I have a field of type `string`, the union type will currently be `string|Foo` right? Where just `string` will be enough?

I guess we can do this as an improvement.

Reply:
Shall we create an issue to track this?

Reply:
Issue to track this https://github.com/ballerina-platform/ballerina-lang/issues/20252